### ISSUE TITLE
Issues when using blocksparse to build simple local block attentions

### ISSUE THREAD

**Issue by xwhan** [facebookresearch/xformers#205]
# ❓ Questions and Help

@blefaudeux @dianaml0 

1. I found the doc [here](https://github.com/facebookresearch/xformers/blob/main/HOWTO.md#blocksparseattention) is a bit confusing: it passed a layout of size [SEQ//BLOCK_SIZE, SEQ//BLOCK_SIZE] but the code comments [here](https://github.com/facebookresearch/xformers/blob/main/xformers/components/attention/blocksparse.py#L49) says the layout should have size [HEAD, SEQ, SEQ]. The layout I passed to build the local block attention is 
```
            blocks = seq_len // block_size
            layout = torch.eye(blocks)
```
Is this the right way to do this?

2. What are the environment requirements for Triton to work? I got this error https://github.com/openai/triton/issues/322  with CUDA 11.1 on A100. I tried to update my cuda toolkit to 11.3, 11.4.2 but both did not work for me

Thanks in advance!

**Comment by blefaudeux** [comment]
Hi @xwhan, thanks for the questions, we need to improve the doc !

1 - **Mask: [H, blocks, blocks] or [blocks, blocks] ?** 
you can do both actually (head dimension or no head dimension). The caveat is that if your mask don't have a head dimension, then xformers will assume that you want to use the same mask across all heads. If your mask has a head dimension, then it can differ as you wish across heads.

2 - * "Invalid image" when trying to use triton, what are the requirements in general ?* 
- HW requirements is a GPU with tensor cores (V100 and newer on the pro side, gtx 1xxx and newer on the consumer side)
- SW requirement should just be a recent cuda, but I've heard problems like yours indeed on some managed envs (@fmassa had that recently). 'nvidia-smi' is not a very good friend, in that it does not always report the right cuda if I remember correctly. Normally you should "just" (sorry for the annoyance) have to do a "module unload cuda" (to make sure that you don't have a trailing cuda runtime around) then "module load cuda/11.5" or whatever you have access to. I hope that helps !

**Comment by blefaudeux** [comment]
and as per your example
`blocks = seq_len // block_size
 layout = torch.eye(blocks)`
yes that works, all heads will use the same pattern (you will have a warning that the mask is extended to all heads) and it will be block diagonal (which is not the same as band diagonal, I'm sure you know but writing this down here if it helps somebody)

**Comment by xwhan** [comment]
nvcc should give me the right version right? 11.4.2 still does not work for me

**Comment by xwhan** [comment]
"Invalid image" resolved after updating to cuda 11.6

**Comment by blefaudeux** [comment]
> "Invalid image" resolved after updating to cuda 11.6

Great, sorry for all the trouble, I'll keep this in mind if sombedoy gets the same issue. I forgot to mention that but I've been using 11.5 without problems
